Fix documentation for {{each}} and {{with}} helpers

YoranBrondsema opened the pull request:
Addresses https://github.com/emberjs/ember.js/issues/10994.
## {{each}} helper

Restored the original documentation from 1.11 with the following changes:
- Updated to block params syntax (e.g. replace `{{#each post in posts}}` with `{{#each posts as |post|}}`
- Removed mentions of `ObjectController` and replace with `Controller`.
- Replaced `{{crossLink}}` helpers with links to relevant documentation.
## {{with}} helper
- Updated to block params syntax
- Expanded the example for the section about the `controller` option
## Comment about documentation of parameters

I still think in both cases the description of the parameters are unclear and I think this holds for all helpers. Like the `{{each}}` helper documents the parameters `name` and `path` but it's not clear how they're used and what they represent.

Instead, would it make sense to document the syntax of the helpers with a sort of grammar? I like how MDN documents the formal syntax of CSS rules. For instance, the formal syntax of the `box-shadow` rule is as follows (https://developer.mozilla.org/en-US/docs/Web/CSS/box-shadow#Formal_syntax):

```
none | [inset? && [ <offset-x> <offset-y> <blur-radius>? <spread-radius>? <color>? ] ]#
```

Using this for helpers you could have something as follows for `{{each}}`:

```
{{each <array> [itemViewClass="<item-view-class>"]? [emptyViewClass="<empty-view-class>"]? [itemController="<item-controller>"]? as |<item>|}}
```

In any case, correct me if I'm wrong but I don't think that in their current format, the documentation of the parameters are particularly useful to most of the Ember.js users.

YoranBrondsema committed:
Fix documentation for {{each}} helper.

YoranBrondsema committed:
Fix documentation for {{with}} helper.

mixonic commented on packages/ember-htmlbars/lib/helpers/each.js:
````diff
@@ -2,6 +2,151 @@ import { get } from "ember-metal/property_get";
+
+  ```handlebars
+  <ul>
+    {{#each developers itemViewClass="person" as |developer|}}
````
We should probably not doc this, as it creates poor performance and will be removed.

rwjblue replied:
It is also deprecated as of 1.13.

YoranBrondsema replied:
OK I removed this part of the documentation.

mixonic commented on packages/ember-htmlbars/lib/helpers/each.js:
````diff
@@ -2,6 +2,151 @@ import { get } from "ember-metal/property_get";
+
+  ```handlebars
+  <ul>
+    {{each developers itemViewClass="person" as |developer|}}
````
OMG, this makes my brain hurt so bad :-)

mixonic commented on packages/ember-htmlbars/lib/helpers/each.js:
````diff
@@ -2,6 +2,151 @@ import { get } from "ember-metal/property_get";
+  ```javascript
+  App.NoPeopleView = Ember.View.extend({
+    tagName: 'li',
+    template: 'No person is available, sorry'
````
Not a valid template

mixonic commented:
@YoranBrondsema this is an improvement. JSCS is complaining about some whitespace in the code doc blocks (you can find the failure locally by running the tests).

Can we remove all mention of setting custom controllers and views? These APIs, in glimmer, are much slower than just using a component in side the each block, for instance. I think this is a good chance flag them in the documentation as legacy.

@rwjblue are the `itemViewClass` and `controller=` APIs going to be deprecated in the 1.13.x cycle?

rwjblue commented:
@mixonic - Anything that opts you into the `LegacyEachView` was deprecated in https://github.com/emberjs/ember.js/pull/11129.

YoranBrondsema committed:
Merge branch 'master' into yb-fix-documentation-for-each-and-with-helpers

YoranBrondsema committed:
Add Vim swp files to .gitignore

YoranBrondsema committed:
Remove documentation for itemViewClass for the {{each}} helper

YoranBrondsema committed:
Make template in documentation for {{each}} valid.

YoranBrondsema committed:
Remove trailing whitespace

YoranBrondsema committed:
Remove the documentation for itemController on {{each}}.

YoranBrondsema committed:
Remove documentation for emptyViewClass for {{each}} helper

YoranBrondsema commented:
OK I removed the documentation for `itemController`, `itemViewClass` and `emptyViewClass` for `{{each}}`.

Also fixed the trailing whitespace.

For `{{with}}`, now I'm not sure whether the `controller` parameter should stay or go?

rwjblue commented:
`controller` should be removed from documentation (it will be deprecated in 1.13 just hasn't happened yet).

YoranBrondsema committed:
Remove documentation for 'controller=' for {{with}} helper

rwjblue commented on .gitignore:
```diff
@@ -39,3 +39,4 @@ publish_to_bower/
 bower_components/
 npm-debug.log
 .ember-cli
+*.swp
```
This should generally go in a global `.gitignore`. See [here](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore) for info...

YoranBrondsema replied:
That makes sense, thanks.

YoranBrondsema committed:
Remove '*.swp' from .gitignore

mixonic commented:
restarted the build

mixonic commented:
@YoranBrondsema thank you!!
